source/MaterialXCore/MaterialNode.cpp

```diff
@@ -0,0 +1,151 @@
+                            {
+                                std::list<NodePtr> newShaderNodes = getShaderNodes(upstreamNode, shaderType, target);
+                                if (!newShaderNodes.empty())
+                                {
+                                    shaderNodes.insert(shaderNodes.end(), newShaderNodes.begin(), newShaderNodes.end());
+                                }
+                            }
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    for (const InputPtr& input : inputs) {
+        const string& inputShader = input->getNodeName();
+        if (!inputShader.empty())
+        {
+            NodePtr shaderNode = parent->getChildOfType<Node>(inputShader);
+            if (shaderNode)
+            {
+                if (!target.empty())
+                {
+                    // Look the shader definition, optionally filtered by target
+                    NodeDefPtr nodeDef = shaderNode->getNodeDef(target);
+                    if (!nodeDef)
+                    {
+                        continue;
+                    }
+                }
+                if (shaderType.empty() || shaderNode->getType() == shaderType)
+                {
+                    shaderNodes.push_back(shaderNode);
+                }
+            }
+        }
+        else
+        {
+            const string& inputGraph = input->getNodeGraphName();
+            if (!inputGraph.empty())
+            {
+                NodeGraphPtr nodeGraph = parent->getChildOfType<NodeGraph>(inputGraph);
+                if (nodeGraph)
+                {
+                    const string& nodeGraphOutput = input->getOutputString();
+                    OutputPtr out = nullptr;
+                    if (!nodeGraphOutput.empty())
+                    {
+                        out = nodeGraph->getOutput(nodeGraphOutput);
+                    }
+                    else
+                    {
+                        std::vector<OutputPtr> outputs = nodeGraph->getOutputs();
+                        if (!outputs.empty())
+                        {
+                            out = outputs[0];
+                        }
+                    }
+                    if (out)
+                    {
+                        if (shaderType.empty() || out->getType() == shaderType)
+                        {
+                            NodePtr upstreamNode = out->getConnectedNode();
+                            if (upstreamNode)
+                            {
+                                shaderNodes.push_back(upstreamNode);
+                            }
+                        }
+                    }
+                }
+            }
+        }
+    }
+    return shaderNodes;
+}
+
+vector<MaterialAssignPtr> getGeometryBindings(NodePtr materialNode, const string& geom)
+{
+    vector<MaterialAssignPtr> matAssigns;
+    for (LookPtr look : materialNode->getDocument()->getLooks())
```

Comment:
In this case no nullptr check is made on the input. Sending in materialNode as a const ref instead would be more safe to make sure it's never null.